### DISCUSSION TITLE
Rework convert to send per-disc file lists to the pipeline

### DISCUSSION THREAD

**Discussion by RollingStar** [beetbox/beets#3917] [Ideas]
I am entering the wild world of gapless audio. Apparently opus needs some hacks to make it work. The encoder requires a series of sequential files, which are joined seamlessly, encoded, then ripped apart without gaps.

https://hydrogenaud.io/index.php?topic=120926.0

https://hydrogenaud.io/index.php?topic=116605.50

Related issue for MP3s, but apparently you can handle those files individually. https://beets.readthedocs.io/en/latest/plugins/convert.html

https://github.com/beetbox/beets/blob/master/beetsplug/convert.py

In beets, `convert` only works on individual files. 

It might be an easy fix to slot in with opts.album already existing? https://github.com/beetbox/beets/blob/master/beetsplug/convert.py#L478

Make the pipe iterate over albums and not items?

https://github.com/beetbox/beets/blob/master/beetsplug/convert.py#L509

**Comment by sampsyo** [comment]
Fascinating. Providing this at least as an option seems like a good idea. It seems like a potential complexity here is that the user's configuration will need to be aware this is going on… that is, you might have to provide _different_ command-line invocations for albums and tracks now? And some conversion tools may not be able to handle multiple tracks at all. So it seems like there is some careful thinking to be done about how to keep the current use cases while adding this capability… preferably while making the built-in ffmpeg configurations "gapless by default," unless that comes with an obvious downside I'm not thinking of.

**Reply by RollingStar** [reply]
It might be too much trouble to treat `-a` as "encode as albums". All for the best, perhaps, because it should really be by discs (-d?). 30+ disc compilations shouldn't be joined into one gapless odyssey. Sometimes disc 2 is a live set or something and the replaygain etc are not going to be the same.

> you might have to provide different command-line invocations for albums and tracks now

Maybe, or maybe not. There's already different "conversions" that can be defined. Converting by discs would be an advanced feature the user could configure appropriately. Most encoders can't handle it so a user would have to set it up specially, or we could provide a default that does this gapless opus thing I want. I don't think we can trash the per-track conversion default.
